### PR TITLE
Link MetadataDatabase to a PACS

### LINKED ISSUE #32: List of avaiblable data
**GIVEN** user loads the application web pages
**WHEN** main page appears
**THEN** a list shall be appear about the data available in the database with the following main parameters:
- Title of Data item
- Unique id of Data item
- Modality of Medical DICOM Data
- Tags
- Last Updated

_Note: this should be reviewed and approved after data model is available_

List of tasks:
- [x] Start to feed MongoDB
- [ ] Define API (depends on data model)
- [x] Create main page in frontend
- [ ] Implement business logic

### PR DESCRIPTION
**Pull request by bat3**
Link MetadataDatabase to a PACS:
To the launching of MetadataDatabase a mirror between mongo data base and a PACS is done.
Every missing Series from the PACS will be created.
Every Series not on the PACS but in mogo DB will be deleted.

### CODE REVIEW

File: business/MetadataDatabase/Services/PacsMirrorServices.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+        public void MirrorPacs()
+        {
+            // Get all series on PACS
+            Task<IEnumerable<QidoSeries>> task = this.pacsService.GetSeries();
+            task.Wait();
+            var pacsSeries = task.Result?.ToDto();
+            // Get all series on DB
+            var dataBaseSeries = this.seriesService.GetAll();
+            // Get SeriesInstanceUID to compaire
```
Comment (cquennouelle) [inline business/MetadataDatabase/Services/PacsMirrorServices.cs]: compare

File: business/MetadataDatabase/Services/PacsMirrorServices.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            IEnumerable<string> pacsSeriesUids = pacsSeries.Select(series => series.SeriesInstanceUID);
+            IEnumerable<string> dataBaseSeriesUids = dataBaseSeries.Select(series => series.SeriesInstanceUID);
+            // Delete into DB all series not in the PACS
+            Console.WriteLine("series Not In Pacs");
```
Comment (cquennouelle) [inline business/MetadataDatabase/Services/PacsMirrorServices.cs]: Write this comment only if some series are not in the PACS. Or at least add a ':'

File: business/MetadataDatabase/Services/PacsMirrorServices.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            IEnumerable<string> seriesUidsNotInPacs = dataBaseSeriesUids.Except(pacsSeriesUids);
+            foreach (string uid in seriesUidsNotInPacs)
+            {
+                Console.WriteLine(uid);
+                IEnumerable<SeriesDto> query = dataBaseSeries.Where(series => series.SeriesInstanceUID == uid);
+                this.seriesService.Delete(query.FirstOrDefault().Id);
+            }
+            // Create into DB all series not in the DB but in PACS
+            Console.WriteLine("series Not In Db");
```
Comment (cquennouelle) [inline business/MetadataDatabase/Services/PacsMirrorServices.cs]: Idem add a ':'

File: business/MetadataDatabase/appsettings.json [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
-    }
-  },
-  "AllowedHosts": "*"
+    "MongoConfiguration": {
+        "DatabaseName": "healthbarbecue",
+        "Servername": "localhost",
+        "Port": "27017"
+    },
+    "PacsConfiguration": {
+        "Host": "localhost",
```
Comment (ptitdam2001) [inline business/MetadataDatabase/appsettings.json:8]: You have to change docker-compose file, you must add environment variables to hb_business_server adding the line:
- PacsConfiguration__Host: hb_pacs_server

For the moment, the production stack is not ok due to the bad servername

File: business/MetadataDatabase/Models/QidoSeries.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        // A number that identifies this Series. (00200011)
+        [JsonPropertyName("00200011")]
+        public DicomIntObject SeriesNumber { get; set; }
+
+        // The number of Composite Object Instances in a Series that match the Series level Query/Retrieve search criteria (00201209)
+        [JsonPropertyName("00201209")]
+        public DicomIntObject NumberOfSeriesRelatedInstances { get; set; }
+
+        // Todo
+        public string GetValueOfDicomTag(DicomTag propertyName)
```
Comment (Nalien) [inline business/MetadataDatabase/Models/QidoSeries.cs]: To move to a service

File: business/MetadataDatabase/Models/QidoSeries.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                        resultValue = dicomNameObject.Value[0].Alphabetic;
+                    }
+                    break;
+                default:
+                    break;
+            }
+            return resultValue;
+        }
+
+        public class DicomTag
```
Comment (Nalien) [inline business/MetadataDatabase/Models/QidoSeries.cs]: move to dedicated file

Comment (Nalien) [reply business/MetadataDatabase/Models/QidoSeries.cs]: Replace this by en enum !
in C#, myEnum.MyEnumeValue123.ToString() returns "MyEnumeValue123"

File: business/MetadataDatabase/Models/QidoSeries.cs [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            public static DicomTag PatientsBirthDate { get { return new DicomTag("PatientsBirthDate"); } }
+            public static DicomTag PatientsSex { get { return new DicomTag("PatientsSex"); } }
+            public static DicomTag StudyInstanceUID { get { return new DicomTag("StudyInstanceUID"); } }
+            public static DicomTag SeriesInstanceUID { get { return new DicomTag("SeriesInstanceUID"); } }
+            public static DicomTag StudyID { get { return new DicomTag("StudyID"); } }
+            public static DicomTag SeriesNumber { get { return new DicomTag("SeriesNumber"); } }
+            public static DicomTag NumberOfSeriesRelatedInstances { get { return new DicomTag("NumberOfSeriesRelatedInstances"); } }
+        }
+    }
+    public class DicomStringObject
```
Comment (Nalien) [inline business/MetadataDatabase/Models/QidoSeries.cs]: move to dedicated files

File: business/MetadataDatabase/Services/IPacsMirrorServices.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿namespace MetadataDatabase.Services
+{
+    public interface IPacsMirrorServices
```
Comment (Nalien) [inline business/MetadataDatabase/Services/IPacsMirrorServices.cs]: rename to IPacsMirrorService

File: business/MetadataDatabase/Services/IPacsMirrorServices.cs [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+    {
+        public void MirrorPacs();
```
Comment (Nalien) [inline business/MetadataDatabase/Services/IPacsMirrorServices.cs]: what is this function doing ?

File: business/MetadataDatabase/Services/PacsMirrorServices.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            Task<IEnumerable<QidoSeries>> task = this.pacsService.GetSeries();
+            task.Wait();
+            var pacsSeries = task.Result?.ToDto();
+            // Get all series on DB
+            var dataBaseSeries = this.seriesService.GetAll();
+            // Get SeriesInstanceUID to compare
+            IEnumerable<string> pacsSeriesUids = pacsSeries.Select(series => series.SeriesInstanceUID);
+            IEnumerable<string> dataBaseSeriesUids = dataBaseSeries.Select(series => series.SeriesInstanceUID);
+            // Delete into DB all series not in the PACS
+            Console.WriteLine("series Not In Pacs:");
```
Comment (Nalien) [inline business/MetadataDatabase/Services/PacsMirrorServices.cs]: remove all Console.WriteLine.
replace by a Log (use a TODO LOG meanwhile)

File: business/MetadataDatabase/Services/PacsServices.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        private static readonly HttpClient client = new HttpClient();
+        private readonly PacsConfiguration settings;
+
+        public PacsServices(IOptions<PacsConfiguration> settings)
+        {
+            this.settings = settings.Value;
+        }
+
+        public async Task<IEnumerable<QidoSeries>> GetSeries()
```
Comment (Nalien) [inline business/MetadataDatabase/Services/PacsServices.cs]: namming convension : Async function should be suffixed by Async
rename to GetSeriesAsync()

File: business/MetadataDatabase/Startup.cs [hunk 12]
```diff
@@ -48,14 +48,14 @@ public void ConfigureServices(IServiceCollection services)
...
-        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, IPacsMirrorServices pacsMirrorServices)
         {
             if (env.IsDevelopment())
             {
                 app.UseDeveloperExceptionPage();
             }
 
-            // app.UseHttpsRedirection();
+            pacsMirrorServices.MirrorPacs();
```
Comment (Nalien) [inline business/MetadataDatabase/Startup.cs]: strictly forbiden in a Startup !!!

**Review by Nalien — CHANGES_REQUESTED** [review]

File: business/MetadataDatabase/Services/IPacsMirrorServices.cs (hunk 8 above)
Comment (bat3) [reply business/MetadataDatabase/Services/IPacsMirrorServices.cs]: Done and other files with "services"

File: business/MetadataDatabase/Startup.cs (hunk 12 above)
Comment (bat3) [reply business/MetadataDatabase/Startup.cs]: Deleted

File: business/MetadataDatabase/Services/PacsServices.cs (hunk 11 above)
Comment (bat3) [reply business/MetadataDatabase/Services/PacsServices.cs]: Done

File: business/MetadataDatabase/Services/PacsMirrorServices.cs (hunk 10 above)
Comment (bat3) [reply business/MetadataDatabase/Services/PacsMirrorServices.cs]: Deleted

File: business/MetadataDatabase/Services/IPacsMirrorServices.cs (hunk 9 above)
Comment (bat3) [reply business/MetadataDatabase/Services/IPacsMirrorServices.cs]: Adds comments

**Review by cquennouelle — DISMISSED** [review]

File: business/MetadataDatabase/Models/QidoSeries.cs (hunk 5 above)
Comment (bat3) [reply business/MetadataDatabase/Models/QidoSeries.cs]: @Nalien, I took into account all your reviews except this one. I don't see why in a service. this method is an help to get properties of this class.

**Review by ptitdam2001 — DISMISSED** [review]

File: client/web/src/components/series/list/SeriesList.tsx [hunk 13]
```diff
@@ -13,44 +13,39 @@ import {
...
-  }
+  useEffect(() => {
+    dispatch({ type: actionTypes.FETCH_ALL_SERIES })
+  }, [dispatch])
 
-  const refresh = () => {
-    axios.get('/api/PacsMirror').then(() => {
-      updateSeriesList();
+  const synchronize = () => {
+    let httpRequest = new HttpRequest({ baseURL: '/api' })
```
Comment (ptitdam2001) [inline client/web/src/components/series/list/SeriesList.tsx]: Bad place, we should move the code :
- First: add new Feature folder Pacs
- first: create one service about Pacs  with your HttpRequest
- Next: call the method into serieSaga => fetchSeries using code below:
```typescript
yield call([pacService, 'fetchMiror'])
....
```

For more information : https://redux-saga.js.org/docs/advanced/SequencingSagas.html
the instruction yield is waiting for the function is terminated to continue the process

Doing with that method, we remove business logic from components

File: client/web/src/components/series/list/SeriesList.tsx [hunk 14]
```diff
@@ -13,44 +13,39 @@ import {
...
-    updateSeriesList();
-  }, [])
-
   return (
     <TableContainer component={Paper}>
       <Grid container justify="flex-end">
         <Button 
           variant="contained"
-          onClick={() => refresh()}
+          onClick={() => synchronize()}
```
Comment (ptitdam2001) [inline client/web/src/components/series/list/SeriesList.tsx:45]: synchronize could be a call of dispatch :+1: 

```typescript
const synchonize = () => {
  dispatch({ type: typePacsAction.DO_A_MIRROR_UPDATE)
}
```

**Review by ptitdam2001 — APPROVED** [review]

**Pull request merged**